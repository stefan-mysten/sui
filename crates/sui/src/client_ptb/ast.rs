// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::fmt;

use move_command_line_common::{
    address::{NumericalAddress, ParsedAddress},
    types::{ParsedFqName, ParsedModuleId, ParsedStructType, ParsedType},
};
use move_core_types::runtime_value::MoveValue;
use sui_types::{base_types::ObjectID, Identifier};

use crate::{error_, sp_};

use super::error::{PTBResult, Span, Spanned};

pub type ParsedProgram = (Program, ProgramMetadata);

// Commands
pub const TRANSFER_OBJECTS: &str = "transfer-objects";
pub const SPLIT_COINS: &str = "split-coins";
pub const MERGE_COINS: &str = "merge-coins";
pub const MAKE_MOVE_VEC: &str = "make-move-vec";
pub const MOVE_CALL: &str = "move-call";
pub const PUBLISH: &str = "publish";
pub const UPGRADE: &str = "upgrade";
pub const ASSIGN: &str = "assign";
pub const PREVIEW: &str = "preview";
pub const WARN_SHADOWS: &str = "warn-shadows";
pub const PICK_GAS_BUDGET: &str = "pick-gas-budget";
pub const GAS_BUDGET: &str = "gas-budget";
pub const SUMMARY: &str = "summary";
pub const GAS_COIN: &str = "gas-coin";
pub const JSON: &str = "json";

// Types
pub const U8: &str = "u8";
pub const U16: &str = "u16";
pub const U32: &str = "u32";
pub const U64: &str = "u64";
pub const U128: &str = "u128";
pub const U256: &str = "u256";

// Keywords
pub const ADDRESS: &str = "address";
pub const BOOL: &str = "bool";
pub const VECTOR: &str = "vector";
pub const SOME: &str = "some";
pub const NONE: &str = "none";
pub const GAS: &str = "gas";

/// A PTB Program consisting of a list of commands and a flag indicating if the preview
/// warn-shadows command was present.
#[derive(Debug, Clone)]
pub struct Program {
    pub commands: Vec<Spanned<ParsedPTBCommand>>,
    // Held outside of metadata since this is used by the PTB builder
    pub warn_shadows_set: bool,
}

/// The `ProgramMetadata` struct holds metadata about a PTB program, such as whether the preview
/// flag was set, json output was set, etc.
#[derive(Debug, Clone)]
pub struct ProgramMetadata {
    pub preview_set: bool,
    pub summary_set: bool,
    pub gas_object_id: Option<Spanned<ObjectID>>,
    pub json_set: bool,
}

/// Types of gas pickers that can be used to pick a gas budget from a list of gas budgets.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GasPicker {
    Max,
    Sum,
}

/// A parsed module access consisting of the address, module name, and function name.
#[derive(Debug, Clone)]
pub struct ModuleAccess {
    pub address: Spanned<ParsedAddress>,
    pub module_name: Spanned<Identifier>,
    pub function_name: Spanned<Identifier>,
}

/// A parsed PTB command consisting of the command and the parsed arguments to the command.
#[derive(Debug, Clone)]
pub enum ParsedPTBCommand {
    TransferObjects(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
    SplitCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
    MergeCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
    MakeMoveVec(Spanned<ParsedType>, Spanned<Vec<Spanned<Argument>>>),
    MoveCall(
        Spanned<ModuleAccess>,
        Option<Spanned<Vec<ParsedType>>>,
        Vec<Spanned<Argument>>,
    ),
    Assign(Spanned<String>, Option<Spanned<Argument>>),
    Publish(Spanned<String>),
    Upgrade(Spanned<String>, Spanned<Argument>),
    PickGasBudget(Spanned<GasPicker>),
    GasBudget(Spanned<u64>),
    WarnShadows,
    Preview,
}

/// An enum representing the parsed arguments of a PTB command.
#[derive(Debug, Clone)]
pub enum Argument {
    Bool(bool),
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    U256(move_core_types::u256::U256),
    Gas,
    Identifier(String),
    VariableAccess(Spanned<String>, Vec<Spanned<String>>),
    Address(NumericalAddress),
    String(String),
    Vector(Vec<Spanned<Argument>>),
    Option(Spanned<Option<Box<Argument>>>),
}

impl Argument {
    /// Resolve an `Argument` into a `MoveValue` if possible. Errors if the `Argument` is not
    /// convertible to a `MoveValue`.
    pub fn to_move_value_opt(&self, loc: Span) -> PTBResult<MoveValue> {
        Ok(match self {
            Argument::Bool(b) => MoveValue::Bool(*b),
            Argument::U8(u) => MoveValue::U8(*u),
            Argument::U16(u) => MoveValue::U16(*u),
            Argument::U32(u) => MoveValue::U32(*u),
            Argument::U64(u) => MoveValue::U64(*u),
            Argument::U128(u) => MoveValue::U128(*u),
            Argument::U256(u) => MoveValue::U256(*u),
            Argument::Address(a) => MoveValue::Address(a.into_inner()),
            Argument::Vector(vs) => MoveValue::Vector(
                vs.iter()
                    .map(|sp_!(loc, v)| v.to_move_value_opt(*loc))
                    .collect::<PTBResult<Vec<_>>>()
                    .map_err(|e| e.with_help(
                            format!("Was unable to parse '{self}' as a pure PTB value. This is most likely because \
                                    the vector contains non-primitive (e.g., object or array) \
                                    values which aren't permitted inside vectors")
                            ))?
            ),
            Argument::String(s) => {
                MoveValue::Vector(s.bytes().map(MoveValue::U8).collect::<Vec<_>>())
            }
            Argument::Option(sp_!(loc, o)) => {
                if let Some(v) = o {
                    let v = v.as_ref().to_move_value_opt(*loc).map_err(|e| e.with_help(
                            format!(
                                "Was unable to parse '{self}' as a pure PTB value. This is most likely because \
                                the option contains a non-primitive (e.g., object or array) \
                                value which isn't permitted inside an option"
                                )
                            ))?;
                    MoveValue::Vector(vec![v])
                } else {
                    MoveValue::Vector(vec![])
                }
            }
            Argument::Identifier(_)
            | Argument::VariableAccess(_, _)
            | Argument::Gas => error_!(loc, "Was unable to convert '{self}' to primitive value (i.e., non-object value)"),
        })
    }
}

impl fmt::Display for Argument {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Argument::Bool(b) => write!(f, "{}", b),
            Argument::U8(u) => write!(f, "{}u8", u),
            Argument::U16(u) => write!(f, "{}u16", u),
            Argument::U32(u) => write!(f, "{}u32", u),
            Argument::U64(u) => write!(f, "{}u64", u),
            Argument::U128(u) => write!(f, "{}u128", u),
            Argument::U256(u) => write!(f, "{}u256", u),
            Argument::Gas => write!(f, "gas"),
            Argument::Identifier(i) => write!(f, "{}", i),
            Argument::VariableAccess(sp_!(_, head), accesses) => {
                write!(f, "{}", head)?;
                for sp_!(_, access) in accesses {
                    write!(f, ".{}", access)?;
                }
                Ok(())
            }
            Argument::Address(a) => write!(f, "@{}", a),
            Argument::String(s) => write!(f, "\"{}\"", s),
            Argument::Vector(v) => {
                write!(f, "vector[")?;
                for (i, sp_!(_, arg)) in v.iter().enumerate() {
                    write!(f, "{}", arg)?;
                    if i != v.len() - 1 {
                        write!(f, ", ")?;
                    }
                }
                write!(f, "]")
            }
            Argument::Option(sp_!(_, o)) => match o {
                Some(v) => write!(f, "some({})", v),
                None => write!(f, "none"),
            },
        }
    }
}

impl fmt::Display for GasPicker {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use GasPicker::*;
        match self {
            Max => write!(f, "max"),
            Sum => write!(f, "sum"),
        }
    }
}

impl fmt::Display for ParsedPTBCommand {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ParsedPTBCommand::TransferObjects(arg, args) => write!(
                f,
                "{TRANSFER_OBJECTS} {} [{}]",
                arg.value,
                args.value
                    .iter()
                    .map(|x| x.value.to_string())
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            ParsedPTBCommand::SplitCoins(arg, args) => write!(
                f,
                "{SPLIT_COINS} {} [{}]",
                arg.value,
                args.value
                    .iter()
                    .map(|x| x.value.to_string())
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            ParsedPTBCommand::MergeCoins(arg, args) => write!(
                f,
                "{MERGE_COINS} {} [{}]",
                arg.value,
                args.value
                    .iter()
                    .map(|x| x.value.to_string())
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            ParsedPTBCommand::Assign(arg, arg_opt) => write!(
                f,
                "{ASSIGN} {}{}",
                arg.value,
                if let Some(arg) = arg_opt {
                    format!(" {}", arg.value)
                } else {
                    "".to_string()
                }
            ),
            ParsedPTBCommand::Publish(s) => write!(f, "{PUBLISH} {}", s.value),
            ParsedPTBCommand::Upgrade(s, a) => write!(f, "{UPGRADE} {} {}", s.value, a.value),
            ParsedPTBCommand::WarnShadows => write!(f, "{WARN_SHADOWS}"),
            ParsedPTBCommand::Preview => write!(f, "{PREVIEW}"),
            ParsedPTBCommand::PickGasBudget(picker) => {
                write!(f, "{PICK_GAS_BUDGET} {}", picker.value)
            }
            ParsedPTBCommand::GasBudget(b) => write!(f, "{GAS_BUDGET} {}", b.value),
            ParsedPTBCommand::MakeMoveVec(ty, args) => write!(
                f,
                "{MAKE_MOVE_VEC} {} [{}]",
                TyDisplay(&ty.value),
                args.value
                    .iter()
                    .map(|x| x.value.to_string())
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            ParsedPTBCommand::MoveCall(
                sp_!(
                    _,
                    ModuleAccess {
                        address,
                        module_name,
                        function_name
                    }
                ),
                tys,
                args,
            ) => {
                let address = match &address.value {
                    ParsedAddress::Named(n) => n.to_string(),
                    ParsedAddress::Numerical(n) => n.to_string(),
                };
                let type_args = match tys {
                    Some(tys) => format!(
                        "<{}>",
                        tys.value
                            .iter()
                            .map(|x| TyDisplay(x).to_string())
                            .collect::<Vec<String>>()
                            .join(", ")
                    ),
                    None => "".to_string(),
                };
                write!(
                    f,
                    "{MOVE_CALL} {}::{}::{}{} {}",
                    address,
                    module_name.value,
                    function_name.value,
                    type_args,
                    args.iter()
                        .map(|x| x.value.to_string())
                        .collect::<Vec<String>>()
                        .join(" ")
                )
            }
        }
    }
}

struct TyDisplay<'a>(&'a ParsedType);

impl<'a> fmt::Display for TyDisplay<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use ParsedType::*;
        match self.0 {
            Address => write!(f, "address"),
            U8 => write!(f, "u8"),
            U64 => write!(f, "u64"),
            U128 => write!(f, "u128"),
            Bool => write!(f, "bool"),
            Vector(ty) => write!(f, "vector<{}>", TyDisplay(ty)),
            Struct(ParsedStructType {
                fq_name:
                    ParsedFqName {
                        module: ParsedModuleId { address, name },
                        name: struct_name,
                    },
                type_args,
            }) => {
                let address = match address {
                    ParsedAddress::Named(n) => n.to_string(),
                    ParsedAddress::Numerical(n) => n.to_string(),
                };
                let ty_str = if type_args.is_empty() {
                    "".to_string()
                } else {
                    format!(
                        "<{}>",
                        type_args
                            .iter()
                            .map(|x| TyDisplay(x).to_string())
                            .collect::<Vec<String>>()
                            .join(", ")
                    )
                };
                write!(f, "{}::{}::{}{}", address, name, struct_name, ty_str)
            }
            U16 => write!(f, "u16"),
            U32 => write!(f, "u32"),
            U256 => write!(f, "u256"),
            Signer => write!(f, "signer"),
        }
    }
}
