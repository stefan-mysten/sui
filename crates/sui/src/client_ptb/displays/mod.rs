// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

mod gas_cost_summary;
mod status;
mod summary;

use crate::{client_ptb::ptb::PTBPreview, sp};
use std::fmt::{Display, Formatter};
use tabled::{
    builder::Builder as TableBuilder,
    settings::{style::HorizontalLine, Panel as TablePanel, Style as TableStyle},
};

pub struct Pretty<'a, T>(pub &'a T);

impl<'a> Display for PTBPreview<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut builder = TableBuilder::default();
        let columns = vec!["command", "from"];
        builder.set_header(columns);
        for sp!(loc, cmd) in &self.program.commands {
            builder.push_record([
                cmd.to_string(),
                loc.file_scope.name.to_string(),
            ]);
        }
        let mut table = builder.build();
        table.with(TablePanel::header("PTB Preview"));
        table.with(TableStyle::rounded().horizontals([
            HorizontalLine::new(1, TableStyle::modern().get_horizontal()),
            HorizontalLine::new(2, TableStyle::modern().get_horizontal()),
            HorizontalLine::new(2, TableStyle::modern().get_horizontal()),
        ]));
        table.with(tabled::settings::style::BorderSpanCorrection);

        write!(f, "{}", table)
    }
}
