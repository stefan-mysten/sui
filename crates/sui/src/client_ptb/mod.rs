// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

pub mod displays;
pub mod error;
pub mod lexer;
pub mod ptb;
pub mod ptb_builder;
pub mod token;
pub mod parser;
pub mod ast;
