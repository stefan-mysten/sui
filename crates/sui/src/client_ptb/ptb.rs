// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use super::ptb_builder::{
    ast::{ParsedProgram, Program},
    errors::{FileTable, PTBError},
};
use crate::client_ptb::{
    displays::Pretty,
    ptb_builder::{build_ptb::PTBBuilder, errors::render_errors, parser::ProgramParser},
};

use anyhow::{anyhow, Error};
use clap::{arg, Args};
use move_core_types::account_address::AccountAddress;
use serde::Serialize;
use shared_crypto::intent::Intent;
use std::collections::{BTreeMap, BTreeSet};
use sui_json_rpc_types::{
    SuiExecutionStatus, SuiTransactionBlockEffectsAPI, SuiTransactionBlockResponseOptions,
};
use sui_keys::keystore::AccountKeystore;
use sui_sdk::{wallet_context::WalletContext, SuiClient};
use sui_types::{
    digests::TransactionDigest,
    gas::GasCostSummary,
    quorum_driver_types::ExecuteTransactionRequestType,
    transaction::{ProgrammableTransaction, Transaction, TransactionData},
};

#[derive(Clone, Debug, Args)]
pub struct PTB {
    #[clap(trailing_var_arg = true, allow_hyphen_values = true)]
    pub args: Vec<String>,
}

pub struct PTBPreview<'a> {
    pub program: &'a Program,
}

#[derive(Serialize)]
pub struct Summary {
    pub digest: TransactionDigest,
    pub status: SuiExecutionStatus,
    pub gas_cost: GasCostSummary,
}

impl PTB {
    /// Parses and executes the PTB with the sender as the current active address
    pub async fn execute(
        self,
        args: Vec<String>,
        context: &mut WalletContext,
    ) -> Result<(), Error> {
        let arg_string = args.join(" ");
        let mut file_table = BTreeMap::new();
        let (program, program_metadata) = match Self::parse_ptb_commands(arg_string, &mut file_table)
        {
            Err(errors) => {
                let suffix = if errors.len() > 1 { "s" } else { "" };
                let rendered = render_errors(&file_table, errors);
                eprintln!("Encountered error{suffix} when parsing PTB:");
                for e in rendered.iter() {
                    eprintln!("{:?}", e);
                }
                anyhow::bail!("Could not build PTB due to previous error{suffix}");
            }
            Ok(parsed) => parsed,
        };

        if program_metadata.preview_set {
            println!("{}", PTBPreview { program: &program });
            return Ok(());
        }

        let client = context.get_client().await?;

        let (ptb, budget) = match Self::build_ptb(program, context, client).await {
            Err(errors) => {
                let suffix = if errors.len() > 1 { "s" } else { "" };
                eprintln!("Encountered error{suffix} when building PTB:");
                let rendered = render_errors(&file_table, errors);
                for e in rendered.iter() {
                    eprintln!("{:?}", e);
                }
                anyhow::bail!("Could not build PTB due to previous error{suffix}");
            }
            Ok(x) => x,
        };

        // get all the metadata needed for executing the PTB: sender, gas, signing tx
        // get sender's address -- active address
        let Some(sender) = context.config.active_address else {
            anyhow::bail!("No active address, cannot execute PTB");
        };

        // find the gas coins if we have no gas coin given
        let coins = if let Some(gas) = program_metadata.gas_object_id {
            context.get_object_ref(gas.value).await?
        } else {
            context
                .gas_for_owner_budget(sender, budget, BTreeSet::new())
                .await?
                .1
                .object_ref()
        };

        // get the gas price
        let gas_price = context
            .get_client()
            .await?
            .read_api()
            .get_reference_gas_price()
            .await?;
        // create the transaction data that will be sent to the network
        let tx_data =
            TransactionData::new_programmable(sender, vec![coins], ptb, budget, gas_price);
        // sign the tx
        let signature =
            context
                .config
                .keystore
                .sign_secure(&sender, &tx_data, Intent::sui_transaction())?;

        // execute the transaction
        let transaction_response = context
            .get_client()
            .await?
            .quorum_driver_api()
            .execute_transaction_block(
                Transaction::from_data(tx_data, vec![signature]),
                SuiTransactionBlockResponseOptions::full_content(),
                Some(ExecuteTransactionRequestType::WaitForLocalExecution),
            )
            .await?;

        if let Some(effects) = transaction_response.effects.as_ref() {
            if effects.status().is_err() {
                return Err(anyhow!(
                    "PTB execution {}. Transaction digest is: {}",
                    Pretty(effects.status()),
                    effects.transaction_digest()
                ));
            }
        }

        let summary = {
            let effects = transaction_response.effects.as_ref().ok_or_else(|| {
                anyhow!("Internal error: no transaction effects after PTB was executed.")
            })?;
            Summary {
                digest: transaction_response.digest,
                status: effects.status().clone(),
                gas_cost: effects.gas_cost_summary().clone(),
            }
        };

        if program_metadata.json_set {
            let json_string = if program_metadata.summary_set {
                serde_json::to_string_pretty(&serde_json::json!(summary))
                    .map_err(|_| anyhow!("Cannot serialize PTB result to json"))?
            } else {
                serde_json::to_string_pretty(&serde_json::json!(transaction_response))
                    .map_err(|_| anyhow!("Cannot serialize PTB result to json"))?
            };
            println!("{}", json_string);
        } else if program_metadata.summary_set {
            println!("{}", Pretty(&summary));
        } else {
            println!("{}", transaction_response);
        }

        Ok(())
    }

    /// Exposed for testing
    pub async fn build_ptb(
        program: Program,
        context: &WalletContext,
        client: SuiClient,
    ) -> Result<(ProgrammableTransaction, u64), Vec<PTBError>> {
        let starting_addresses = context
            .config
            .keystore
            .addresses_with_alias()
            .into_iter()
            .map(|(sa, alias)| (alias.alias.clone(), AccountAddress::from(*sa)))
            .collect();
        let builder = PTBBuilder::new(starting_addresses, client.read_api());
        builder.build(program).await
    }

    /// Exposed for testing
    pub fn parse_ptb_commands(
        arg_string: String,
        file_table: &mut FileTable,
    ) -> Result<ParsedProgram, Vec<PTBError>> {
        ProgramParser::new(arg_string, file_table)
            .map_err(|e| vec![e])?
            .parse()
    }
}

fn _ptb_description() -> clap::Command {
    clap::Command::new("ptb")
        .about("Build, preview, and execute programmable transaction blocks.")
        .arg(arg!(
            --file <FILE>
                "Path to a file containing transactions to include in this PTB."
        ))
        .arg(arg!(
            --assign <ASSIGN>...
                "Assign a value to use later in the PTB. If only a name is supplied, the result of \
                 the last transaction is binded to that name. If a name and value are \
                 supplied, then the name is binded to that value."
        ))
        .arg(arg!(
            --gas <ID> ...
                "The object ID of the gas coin to use."
        ))
        .arg(arg!(
            --"gas-budget" <MIST>
                "The gas budget for the transaction, in MIST."
        ))
        .arg(arg!(
            --"make-move-vec" <MAKE_MOVE_VEC>
            r#"Given n-values of the same type, it constructs a vector. For non objects or an empty vector, the type tag must be specified: --make-move-vec "<u64>" "[1]" "#
        ))
    //     #[clap(long, num_args(1..))]
    //     make_move_vec: Vec<String>,
    //     /// Merge N coins into the provided coin: --merge-coins into_coin "[coin1,coin2,coin3]"
    //     #[clap(long, num_args(1..))]
    //     merge_coins: Vec<String>,
    //     /// Make a move call to a function
    //     #[clap(long, num_args(1..))]
    //     move_call: Vec<String>,
    //     /// Split the coin into N coins as per the given amount.
    //     /// On zsh, the vector needs to be given in quotes: --split-coins coin_to_split "[amount1,amount2]"
    //     #[clap(long, num_args(1..))]
    //     split_coins: Vec<String>,
    //     /// Transfer objects to the address. E.g., --transfer-objects to_address "[obj1, obj2]"
    //     #[clap(long, num_args(1..))]
    //     transfer_objects: Vec<String>,
    //     /// Publish the move package. It takes as input the folder where the package exists.
    //     #[clap(long, num_args(1..))]
    //     publish: Vec<String>,
    //     /// Upgrade the move package. It takes as input the folder where the package exists.
    //     #[clap(long, num_args(1..))]
    //     upgrade: Vec<String>,
    //     /// Preview the PTB instead of executing it
    //     #[clap(long)]
    //     preview: bool,
    //     /// Enable shadown warning when including other PTB files.
    //     /// Off by default.
    //     #[clap(long)]
    //     warn_shadows: bool,
    //     /// Pick gas budget strategy if multiple gas-budgets are provided.
    //     #[clap(long)]
    //     pick_gas_budget: Option<PTBGas>,
    //     /// Show only a short summary (digest, execution status, gas cost).
    //     /// Do not use this flag when you need all the transaction data and the execution effects.
    //     #[clap(long)]
    //     summary: bool,
}
