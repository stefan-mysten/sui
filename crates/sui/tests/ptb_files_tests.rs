// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeMap, path::Path};
use sui::client_ptb::{ptb::PTB, ptb_builder::errors::render_errors};
use sui_types::transaction::{CallArg, ObjectArg};
use test_cluster::TestClusterBuilder;

const TEST_DIR: &str = "tests";

#[cfg(not(msim))]
#[tokio::main]
async fn test_ptb_files(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    use sui::client_ptb::ptb::PTBPreview;

    let _ = miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .color(false)
                .width(80)
                .build(),
        )
    }));

    let fname = || path.file_name().unwrap().to_string_lossy().to_string();
    let command = format!("--file '{}'", path.to_string_lossy());
    let mut file_table = BTreeMap::new();

    // Parsing
    let program = PTB::parse_ptb_commands(command, &mut file_table);
    let (program, program_meta) = match program {
        Ok(program) => program,
        Err(errors) => {
            let rendered = render_errors(&file_table, errors);
            let mut results = vec![];
            results.push(" === ERRORS AFTER PARSING INPUT COMMANDS === ".to_string());
            for e in rendered.iter() {
                results.push(format!("{:?}", e));
            }
            insta::assert_display_snapshot!(fname(), results.join("\n"));
            return Ok(());
        }
    };

    // Preview (This is based on the parsed commands).
    let mut results = vec![];
    results.push(" === PREVIEW === ".to_string());
    results.push(format!("{}", PTBPreview { program: &program }));

    results.push(" === PROGRAM META === ".to_string());
    results.push(format!(
        "preview: {}\nsummary: {}\ngas_object: {}\njson: {}",
        program_meta.preview_set,
        program_meta.summary_set,
        program_meta
            .gas_object_id
            .map(|x| x.value.to_string())
            .unwrap_or("none".to_string()),
        program_meta.json_set
    ));

    // === BUILD PTB ===
    let test_cluster = TestClusterBuilder::new().build().await;
    let context = test_cluster.wallet;
    let client = context.get_client().await?;

    let built_ptb = PTB::build_ptb(program, &context, client)
        .await;

    if let Ok(ref ptb) = built_ptb {
        results.push(" === BUILT PTB === ".to_string());
        for (i, ca) in ptb.0.inputs.iter().enumerate() {
            results.push(format!("Input {}: {}", i, stable_call_arg_display(ca)));
        }
        for (i, c) in ptb.0.commands.iter().enumerate() {
            results.push(format!("Command {}: {}", i, c));
        }
    }

    // === BUILDING PTB ERRORS ===
    if let Err(e) = built_ptb {
        let rendered = render_errors(&file_table, e);

        results.push(" === BUILDING PTB ERRORS === ".to_string());
        for e in rendered.iter() {
            results.push(format!("{:?}", e));
        }
    }

    // === FINALLY DO THE ASSERTION ===
    insta::assert_display_snapshot!(fname(), results.join("\n"));

    Ok(())
}

fn stable_call_arg_display(ca: &CallArg) -> String {
    match ca {
        CallArg::Pure(v) => format!("Pure({:?})", v),
        CallArg::Object(oa) => match oa {
            ObjectArg::ImmOrOwnedObject(_) => "ImmutableOrOwnedObject".to_string(),
            ObjectArg::SharedObject { mutable, .. } => {
                format!("SharedObject(mutable: {})", mutable)
            }
            ObjectArg::Receiving(_) => "Receiving".to_string(),
        },
    }
}

datatest_stable::harness!(test_ptb_files, TEST_DIR, r".*\.ptb$",);
