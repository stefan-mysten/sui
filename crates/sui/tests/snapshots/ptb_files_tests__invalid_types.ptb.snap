---
source: crates/sui/tests/ptb_files_tests.rs
expression: "results.join(\"\\n\")"
---
 === ERRORS AFTER PARSING INPUT COMMANDS === 
  × Error at command 0 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[3:12]
 2 │ # Out-of-bounds integers
 3 │ --assign a 0xfffu8
   ·            ───┬───
   ·               ╰── number too large to fit in target type
 4 │ --assign b 0xffffffffu16
   ╰────

  × Error at command 1 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[4:12]
 3 │ --assign a 0xfffu8
 4 │ --assign b 0xffffffffu16
   ·            ──────┬──────
   ·                  ╰── number too large to fit in target type
 5 │ --assign c 0xfffffffffffffu32
   ╰────

  × Error at command 2 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[5:12]
 4 │ --assign b 0xffffffffu16
 5 │ --assign c 0xfffffffffffffu32
   ·            ─────────┬────────
   ·                     ╰── number too large to fit in target type
 6 │ --assign d 0xffffffffffffffffffffffffffffffu64
   ╰────

  × Error at command 3 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[6:12]
 5 │ --assign c 0xfffffffffffffu32
 6 │ --assign d 0xffffffffffffffffffffffffffffffu64
   ·            ─────────────────┬─────────────────
   ·                             ╰── number too large to fit in target type
 7 │ --assign e 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffu128
   ╰────

  × Error at command 4 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[7:12]
 6 │ --assign d 0xffffffffffffffffffffffffffffffu64
 7 │ --assign e 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffu128
   ·            ────────────────────────────────┬───────────────────────────────
   ·                                            ╰── number too large to fit in target type
 8 │ --assign f 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffu256
   ╰────

  × Error at command 5 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
   ╭─[8:12]
 7 │ --assign e 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffu128
 8 │ --assign f 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffu256
   ·            ───────────────────────────────────────────┬──────────────────────────────────────────
   ·                                                       ╰── Invalid string length
 9 │ 
   ╰────

  × Error at command 6 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
    ╭─[11:17]
 10 │ # invalid integer types 
 11 │ --make-move-vec <string> ['foo']
    ·                 ────┬───
    ·                     ╰── expected token ::, got >
 12 │ --make-move-vec <String> ['foo']
    ╰────

  × Error at command 7 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
    ╭─[12:17]
 11 │ --make-move-vec <string> ['foo']
 12 │ --make-move-vec <String> ['foo']
    ·                 ────┬───
    ·                     ╰── expected token ::, got >
 13 │ --make-move-vec <Bool> [true, false]
    ╰────

  × Error at command 8 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
    ╭─[13:17]
 12 │ --make-move-vec <String> ['foo']
 13 │ --make-move-vec <Bool> [true, false]
    ·                 ───┬──
    ·                    ╰── expected token ::, got >
 14 │ --make-move-vec <i8> [0]
    ╰────

  × Error at command 9 in PTB file 'tests/ptb_files/parsing/invalid_types.ptb'
    ╭─[14:17]
 13 │ --make-move-vec <Bool> [true, false]
 14 │ --make-move-vec <i8> [0]
    ·                 ──┬─
    ·                   ╰── expected token ::, got >
 15 │ --make-move-vec <i16> [0]
    ╰────

  × Error at command 10 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[15:17]
 14 │ --make-move-vec <i8> [0]
 15 │ --make-move-vec <i16> [0]
    ·                 ──┬──
    ·                   ╰── expected token ::, got >
 16 │ --make-move-vec <i32> [0]
    ╰────

  × Error at command 11 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[16:17]
 15 │ --make-move-vec <i16> [0]
 16 │ --make-move-vec <i32> [0]
    ·                 ──┬──
    ·                   ╰── expected token ::, got >
 17 │ --make-move-vec <i64> [0]
    ╰────

  × Error at command 12 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[17:17]
 16 │ --make-move-vec <i32> [0]
 17 │ --make-move-vec <i64> [0]
    ·                 ──┬──
    ·                   ╰── expected token ::, got >
 18 │ --make-move-vec <i128> [0]
    ╰────

  × Error at command 13 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[18:17]
 17 │ --make-move-vec <i64> [0]
 18 │ --make-move-vec <i128> [0]
    ·                 ───┬──
    ·                    ╰── expected token ::, got >
 19 │ --make-move-vec <i256> [0]
    ╰────

  × Error at command 14 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[19:17]
 18 │ --make-move-vec <i128> [0]
 19 │ --make-move-vec <i256> [0]
    ·                 ───┬──
    ·                    ╰── expected token ::, got >
 20 │ --make-move-vec <Vector<u8>> [vector[0]]
    ╰────

  × Error at command 15 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[20:17]
 19 │ --make-move-vec <i256> [0]
 20 │ --make-move-vec <Vector<u8>> [vector[0]]
    ·                 ──────┬─────
    ·                       ╰── expected token ::, got <
 21 │ --make-move-vec <vector<U256>> [vector[0]]
    ╰────

  × Error at command 16 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[21:17]
 20 │ --make-move-vec <Vector<u8>> [vector[0]]
 21 │ --make-move-vec <vector<U256>> [vector[0]]
    ·                 ───────┬──────
    ·                        ╰── expected token ::, got >
 22 │ --make-move-vec <vector<i256>> [vector[0]]
    ╰────

  × Error at command 17 in PTB file 'tests/ptb_files/parsing/
  │ invalid_types.ptb'
    ╭─[22:17]
 21 │ --make-move-vec <vector<U256>> [vector[0]]
 22 │ --make-move-vec <vector<i256>> [vector[0]]
    ·                 ───────┬──────
    ·                        ╰── expected token ::, got >
    ╰────

